feat(git)!: split off git_tags module from git_commit

saecki opened the pull request:
This separates the git tag segment from the `git_commit` module to a separate `git_tag` module as suggested in #2282.
It also fixes the original issue by taking most of the code from #2102.

#### Description

__Changes to `git_commit` module__
- remove `tag_symbol` variable
- remove `tag_max_candidates` variable
- remove `tag_disabled` variable
- remove `tag` segment

__Add `git_tag` module__
config options:
- `format`
- `symbol` used by default as prefix for tags
- `separator` used as argument to join list of tags
- `style`
- `disabled`

segments:
- `tags`
- `symbol`
- `style`

#### Motivation and Context
This fixes the issue where only annotated tags were shown, and lightweight tags were ignored.

It might also be useful to only show the git tag without the hash.

#### How Has This Been Tested?
- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
    - Added tests to the git_tag module, and run the test suit (the `git_state` test was failing before already, Edit: this was related to my git config).
    - Run `module git_tag` locally in multiple git repositories
    - Run `eval $(starship init zsh)` and test in multiple git repositories.
- [ ] I have tested using **Windows**

#### Checklist:
Do I need to run some command to generate the localized docs?
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

saecki committed:
feat(git)!: split off git_tags module from git_commit

Saecki commented:
Rebased onto master and formatted docs

davidkna reviewed (commented):
Please fix the merge conflicts, so this can be merged, and add the module to the vcs-module format string.

saecki commented:
I think this is ready for review. I didn't find any instructions to update the translated documentation, is this handled by the CI?

saecki commented on src/modules/git_tags.rs:
```diff
@@ -0,0 +1,189 @@
+    let tag_names = all_tags(repo.open())?;
+
+    if tag_names.is_empty() {
+        return Some(module);
```
This could probably just return `None` right?

saecki commented on src/configs/git_tags.rs:
```diff
@@ -0,0 +1,30 @@
+    pub symbol: &'a str,
+    pub separator: &'a str,
+    pub style: &'a str,
+    pub max_candidates: usize,
```
This is outdated

saecki commented:
The windows test failure seems a little spurious 🤔

saecki commented:
Ok, I've renamed the tests to clear up which ones test annotated and which ones test lightweight tags, and added another test case for the mixed case.

This also changes the sorting of all tags to be lexicographic, since this can be done consistently across lightweight and annotated tags, and doesn't require additional traversal of the commit graph. So this should even be more performant.

One thing I'm not sure about is whether we still need to increase the cache size after these changes, which was originally added in #4277. Maybe this is worth looking into...

saecki commented:
This is ready for another round of review.

<sub>(Does this trigger the bot?)</sub>

davidkna reviewed (changes requested):
Please preserve existing options and avoid adding new options as part of this PR.

saecki commented:
> Please preserve existing options and avoid adding new options as part of this PR.

Which options would you like to preserve?
Because the split of the module is intentional the `tag_symbol`, `tag_disabled` and `tag_max_candidates` options of `git_commit` don't really make sense anymore.
A `max_candidates` option for the tags doesn't really apply anymore, because I've changed the way the tags are read. And from what I understand it didn't directly limit how many tags were displayed, but how many candidates were considered when looking for a tag, since only one tag was displayed anyway. We could add another `max_tags` option that limits the number of displayed tags, but that would go against adding more options.

The only "real" new option is the `separator` option of the `git_tags` module, which I think is reasonable to make this configurable. But I can remove it if you want and hard-code it to a space.

davidkna commented:
I haven't seen any demand for displaying multiple tags yet, and `max_candiates` technically allows for interesting more git-describe like display of tags (not sure how much this is used in practice). Displaying all kinds of tags should already be supported.

Besides `max_candiates` I fell like `git_tag` could benefit from `detached_only` for full feature parity with the previous version.

saecki commented:
> I haven't seen any demand for displaying multiple tags yet,

I think it seems quite logical to display multiple tags if a commit is tagged multiple times.

> and `max_candiates` technically allows for interesting more git-describe like display of tags (not sure how much this is used in practice).

Hmm, I'm honestly not sure the git describe behavior is that desirable, and it also seems a bit weird to limit the description only to tags. The `first_parent` option that is set also further limits what `git_describe` can do. 
A `git_describe` module that tries to describe a detached HEAD commit using all available `refs` could be nice for that. Then the `--candidates` (`max_candidates`) flag is would make sense again to limit the computational complexity.
So I think simplifying the `git_tags` module code to what I've done makes sense, and is what is actually expected from a module named that way. 
What do you think?

> Displaying all kinds of tags should already be supported.

I think this wasn't the case when I originally opened the PR, but has since been fixed in #4632. I've updated the PR description.

> Besides `max_candiates` I fell like `git_tag` could benefit from `detached_only` for full feature parity with the previous version.

I'll add that as an option :+1: and resolve the conflicts.

saecki committed:
feat(git_tags): [WIP] add `only_detached` option

injust commented:
Just mentioning that there has been interest in previous issues (e.g. https://github.com/starship/starship/issues/2102#issue-782245583 and https://github.com/starship/starship/issues/2282#issuecomment-1011490038) to hide the commit hash when displaying a tag.
